pkg/apis/admissionregistration/validation/validation.go

```diff
@@ -514,3 +517,145 @@ func ValidateMutatingWebhookConfigurationUpdate(newC, oldC *admissionregistratio
+	}
+	return allErrors
+}
+
+func validateParamSource(ps *admissionregistration.ParamSource, fldPath *field.Path) field.ErrorList {
+	var allErrors field.ErrorList
+	if ps.APIGroup == "" {
+		allErrors = append(allErrors, field.Required(fldPath.Child("apiGroup"), ""))
+	}
+	if ps.APIVersion == "" {
+		allErrors = append(allErrors, field.Required(fldPath.Child("apiVersion"), ""))
+	}
+	if ps.APIKind == "" {
+		allErrors = append(allErrors, field.Required(fldPath.Child("apiKind"), ""))
+	}
+	return allErrors
+}
+
+func validateMatchResources(mc *admissionregistration.MatchResources, fldPath *field.Path) field.ErrorList {
+	var allErrors field.ErrorList
+	if mc.MatchPolicy != nil && !supportedMatchPolicies.Has(string(*mc.MatchPolicy)) {
+		allErrors = append(allErrors, field.NotSupported(fldPath.Child("matchPolicy"), *mc.MatchPolicy, supportedMatchPolicies.List()))
+	}
+	for i, namespace := range mc.Namespaces {
+		for _, msg := range genericvalidation.ValidateNamespaceName(namespace, false) {
+			allErrors = append(allErrors, field.Invalid(fldPath.Child("namespaces").Index(i), namespace, msg))
+		}
+	}
+	for i, namespace := range mc.ExcludeNamespaces {
+		for _, msg := range genericvalidation.ValidateNamespaceName(namespace, false) {
+			allErrors = append(allErrors, field.Invalid(fldPath.Child("excludeNamespaces").Index(i), namespace, msg))
+		}
+	}
+	if mc.NamespaceSelector != nil {
+		allErrors = append(allErrors, metav1validation.ValidateLabelSelector(mc.NamespaceSelector, fldPath.Child("namespaceSelector"))...)
+	}
+
+	if mc.LabelSelector != nil {
+		allErrors = append(allErrors, metav1validation.ValidateLabelSelector(mc.LabelSelector, fldPath.Child("labelSelector"))...)
+	}
+
+	for i, rule := range mc.ResourceRules {
+		allErrors = append(allErrors, validateRuleWithOperations(&rule, fldPath.Child("resourceRules").Index(i))...)
+	}
+
+	for i, rule := range mc.ExcludeResourceRules {
+		allErrors = append(allErrors, validateRuleWithOperations(&rule, fldPath.Child("excludeResourceRules").Index(i))...)
+	}
+
+	return allErrors
+}
+
+func validateValidation(v *admissionregistration.Validation, fldPath *field.Path) field.ErrorList {
+	var allErrors field.ErrorList
+	if v.Name == "" {
+		allErrors = append(allErrors, field.Required(fldPath.Child("name"), ""))
+	}
+	allErrors = append(allErrors, utilvalidation.IsFullyQualifiedName(fldPath.Child("name"), v.Name)...)
+	trimmedExpression := strings.TrimSpace(v.Expression)
+	trimmedMsg := strings.TrimSpace(v.Message)
+	trimmedExpressMsg := strings.TrimSpace(v.ExpressMessage)
+	if len(trimmedExpression) == 0 {
+		allErrors = append(allErrors, field.Required(fldPath.Child("expression"), "expression is not specified"))
+	} else if len(v.ExpressMessage) > 0 && len(trimmedExpressMsg) == 0 {
+		allErrors = append(allErrors, field.Invalid(fldPath.Child("expressMessage"), v.ExpressMessage, "ExpressMessage must be non-empty if specified"))
```

Comment:
Don't repeat field names already stated in the field path in the error string. This applies to all error messages (except ondes like on line 617 that include child field names).

Reply:
This is following the behavior of invalid error which include the value of the field.

Reply:
Including the value is fine. I'm suggesting replacing "ExpressMessage must be non-empty if specified" with "must be non-empty if specified""